compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUPromoteMatmulToFitMMA.cpp

```diff
@@ -0,0 +1,167 @@
+    }
+
+    auto options =
+        linalg::LinalgPaddingOptions()
+            .setPaddingDimensions(paddingDims)
+            .setPaddingValues(paddingValueAttributes)
+            .setPadToMultipleOf(padToMultipleOf)
+            .setPackPaddings(packPaddings)
+            .setCopyBackOp(linalg::LinalgPaddingOptions::CopyBackOp::None);
+
+    FailureOr<linalg::LinalgOp> result =
+        linalg::padAndHoistLinalgOp(rewriter, op, options);
+    if (failed(result)) {
+      LLVM_DEBUG(llvm::dbgs() << "failed to pad op " << op << "\n");
+    }
+  }
+
+  void runOnOperation() override {
+    MLIRContext *ctx = &getContext();
+    auto funcOp = getOperation();
+
+    // Preserve the innermost tensor.pad ops (i.e., pad for reduction dims), so
+    // we can kick canonicalization patterns to fold outer tensor.pad ops away.
+    bool nofold = false;
+    utils::IteratorType targetIterType = utils::IteratorType::parallel;
+    switch (targetDimensions) {
+    case LLVMGPUMatmulPadOption::ParallelDims:
+      LLVM_DEBUG(llvm::dbgs() << "padding parallel dims\n");
+      targetIterType = utils::IteratorType::parallel;
+      nofold = false;
+      break;
+    case LLVMGPUMatmulPadOption::ReductionDims:
+      LLVM_DEBUG(llvm::dbgs() << "padding reduction dims\n");
+      targetIterType = utils::IteratorType::reduction;
+      nofold = true;
+      break;
+    default: // Unreachable.
+      assert(false);
+      break;
+    };
+
+    SmallVector<linalg::LinalgOp> candidates;
+    funcOp->walk([&](linalg::LinalgOp op) {
+      if (linalg::isaContractionOpInterface(op)) {
+        candidates.push_back(op);
+      }
+    });
+
+    IRRewriter rewriter(ctx);
+    for (auto op : candidates) {
+      pad(rewriter, op, targetIterType, nofold);
+    }
+
+    {
+      RewritePatternSet patterns(ctx);
+      linalg::populateSwapExtractSliceWithFillPatterns(patterns);
+      linalg::FillOp::getCanonicalizationPatterns(patterns, ctx);
+      memref::populateResolveRankedShapedTypeResultDimsPatterns(patterns);
+      ctx->getLoadedDialect<tensor::TensorDialect>()
+          ->getCanonicalizationPatterns(patterns);
+      tensor::PadOp::getCanonicalizationPatterns(patterns, ctx);
+      if (failed(applyPatternsAndFoldGreedily(funcOp, std::move(patterns)))) {
+        LLVM_DEBUG(llvm::dbgs() << "----- cleanup failed -----\n");
+        return signalPassFailure();
+      }
+    }
+
+    // XXX(hanchung): This is needed for pad op fusion, which will remove
+    // outer pad ops. I.e., it mainly wants to remove first pad op in the
+    // pad->extract_slice->pad chain, while the canonicalization pattern can
+    // only recognize slice->pad->slice->pad.
+    {
+      SmallVector<tensor::PadOp> padOps;
+      funcOp.walk([&](tensor::PadOp op) { padOps.push_back(op); });
+      for (auto op : padOps) {
+        auto src =
+            op.getSource().getDefiningOp<IREE::Flow::DispatchTensorLoadOp>();
+        if (!src) {
```

Comment:
Since this is just for a `pad->extract_slice->pad` chain, should we check here that the consumer of the pad is a `slice->pad` before adding this?

Reply:
I think it is okay because it is extracting the whole tensor, and will (probably?) get folded away even if there is no `slice->pad` op after.

Reply:
Yeah, I think it is okay, but it's good to do it safely in the fragile pattern. I'll add the check.

Reply:
Yeah even though we're expecting it to fold away, given the fragility of the pattern being conservative seems like a good compromise. Adding the check would be nice, but improving the upstream canonicalization would be best.